Include/internal/pycore_lock.h

```diff
@@ -213,6 +213,41 @@ _PyOnceFlag_CallOnce(_PyOnceFlag *flag, _Py_once_fn_t *fn, void *arg)
     return _PyOnceFlag_CallOnceSlow(flag, fn, arg);
 }
 
+// A readers-writer (RW) lock. The lock supports multiple concurrent readers or
+// a single writer. The lock is write-preferring: if a writer is waiting, then
+// new readers will be blocked. This avoids starvation of writers.
```

Comment:
How far do we want to go with prioritizing waiting writers?

IIUC, currently there's a race with waiting readers and writers once they are all woken up by `_PyParkingLot_UnparkAll()`.  To see what I mean, let's step through the interaction between a handful of threads sharing a `_PyRWMutex`:

1. writer A takes the lock
2. reader B is parked
3. writer C is parked
4. reader D is parked
5. writer A releases the lock
6. the parked threads wake up and take the next iteration in their respective loops (in `_PyRWMutex_RLock()` or `_PyRWMutex_Lock()`)

Since we clear `_Py_HAS_PARKED` at the beginning of `_PyRWMutex_Unlock()`, *before* the `_PyParkingLot_UnparkAll()` call, any one of the following could happen:

<details>
<summary>(expand)</summary>

* both readers beat the writer
   7. readers B and D take the lock
   8. writer C is parked again
   9. readers B and D release the lock
   10. writer C takes the lock
   11. writer C releases the lock
* one reader beats the writer
   7. reader B or D takes the lock
   8. writer C is parked again
   9. the other reader takes the lock too
   10. readers B and D release the lock
   11. writer C takes the lock
   12. writer C releases the lock
* the writer beats the readers
   7. writer C takes the lock
   8. readers B and D are parked again
   9. writer C releases the lock
   10. readers B and D take the lock
   11. readers B and D release the lock

</details>

Thus, in 2 of 3 cases we fail to prioritize a waiting writer over a waiting reader.

### Possible Solutions

I expect the situation would be slightly better if we were to set `_Py_HAS_PARKED` right after the `rwmutex_set_parked_and_wait()` call in `_PyRWMutex_Lock()`.  There would still be a small race with readers, but there's a better chance a waiting writer gets prioritized.

Another idea is to somehow put the readers in a different parking lot (or bucket) than the writers and unpark waiting writers first.

<details>
<summary>One more idea:</summary>

Perhaps there's a way we can be thorough about prioritizing waiting writers.  Let's revisit the steps we took above, re-using steps 1-6.

Here's how it would look if we cleared `_Py_HAS_PARKED` later than we do now: at the point where we don't have any waiting writers left.  That would looks like one of the following (depending on which thread wins:

* both readers beat the writer
   7. readers B and D are parked again (the `_Py_HAS_PARKED` branch)
   8. writer C takes the lock
   9. writer C releases the lock
   10. `_Py_HAS_PARKED` is cleared
   11. readers B and D take the lock
* one reader beats the writer
   7. reader B or D is parked again (the `_Py_HAS_PARKED` branch)
   8. writer C takes the lock
   9. the other reader is parked again
   10. writer C releases the lock
   11. `_Py_HAS_PARKED` is cleared
   12. readers B and D take the lock
* the writer beats the readers
   7. writer C takes the lock
   8. readers B and D are parked again
   9. writer C releases the lock
   10. `_Py_HAS_PARKED` is cleared
   11. readers B and D take the lock

Step 10 (or 11 in the second scenario) is the one that facilitates always prioritizing waiting writers.  To do so, we'd clear `_Py_HAS_PARKED` later than we do now and only if there were no other writers trying to take the lock:

1. also in `_PyRWMutex_Unlock()` but at the end, after the `_PyParkingLot_UnparkAll()` call, and only if there are no other writers trying to take the lock
2. in `_Py_RWMutex_Lock()` in the success case right before returning

In either case, it would require keeping a count of writers, like we do for readers (and checking that count in this branch).

</details>

Reply:
I don't think we want to go any further with prioritizing writers. The goal was narrow: to avoid writer starvation when there are a lot of readers. If we were to further optimize this, we'd probably want to do the opposite: handoff to waiting readers in `_PyRWMutex_Unlock()` so they get a turn too (that's what Go's sync.RWMutex does, for example).

I don't think this is worth optimizing further. For some context, I didn't use this at all in the nogil forks and I'm not sure there will be other uses of this API. That's one of the reasons for "optimizing for simplicity of the implementation".